juju/juju
# [JUJU-2668] Add s3 caller worker, use s3caller to download charms

~_Note:_ This is the continuation of the work done in https://github.com/juju/juju/pull/15212 (it *must* be merged after that PR).~

In this PR the charm downloader is done using a S3 client, more specifically we are going to use the `aws-sdk-go/v2` S3 client. This client is implemented as a new worker `s3Caller`, in order for it to be restarted when the apiserver address changes for some reason. It has a dependency on the `apiCaller` worker, from which it retrieves the apiserver address and the httpclient used to contact it, which is then injected into the aws s3 client.

In this PR the s3 charms downloader only aims at replacing the current downloader which requests the current `/charms` REST API. The generalization of this new downloader to be able to download from other sources (AWS, minio?) needs to be discussed and later implemented.

## Checklist




## QA steps

```sh
$ juju bootstrap localhost ctrl-local
$ juju add-model m
$ juju deploy prometheus2
```
This should work, since both the juju-controller and the prometheus2 charms will be downloaded (from the new S3 endpoint).

~_Note:_ https://github.com/juju/juju/pull/15212 PR must be merged and this branch rebased for this QA to work, since the endpoint is not implemented here.~
